File: lightbite-app/lib/features/auth/domain/repositories/auth_repository.dart
import 'package:dartz/dartz.dart';
import 'package:lightbite_app/core/domain/entities/auth_user.dart';
import 'package:lightbite_app/core/errors/failures.dart';

abstract class AuthRepository {
  Future<AuthUser?> checkAuth();
  Future<Either<Failure, AuthUser>> login(String email, String password);
  Future<Either<Failure, AuthUser>> register(String name, String email, String password, String role);
  Future<void> logout();
}

File: lightbite-kmp/shared/src/commonMain/kotlin/com/lightbite/features/auth/domain/AuthRepository.kt
package com.lightbite.shared.features.auth.domain

import com.lightbite.shared.core.errors.AppResult

/**
 * Repository contract for authentication.
 *
 * Implementation handles:
 *   - SecureStorage for token persistence
 *   - RemoteDataSource for API calls
 *   - Token refresh with queue-safe single-flight pattern
 */
interface AuthRepository {
    /** Check stored tokens and return current auth state. */
    suspend fun checkAuth(): AuthState

    /** Log in with email/password. Persists tokens on success. */
    suspend fun login(credentials: LoginCredentials): AppResult<User>

    /** Register a new account. Persists tokens on success. */
    suspend fun register(credentials: RegisterCredentials): AppResult<User>

    /** Log out. Clears stored tokens. */
    suspend fun logout()

    /** Refresh the access token. Returns new tokens or fails. */
    suspend fun refreshToken(): AppResult<AuthTokens>

    /** Get the current access token (for the API interceptor). */
    suspend fun getAccessToken(): String?

    /** Get the current refresh token. */
    suspend fun getRefreshToken(): String?
}
